tidy3d/components/geometry/base.py

```diff
@@ -1795,6 +1952,320 @@ def _surface_area(self, bounds: Bound) -> float:
+        )
+
+    def intersections_tilted_plane(
+        self, normal: Coordinate, origin: Coordinate, to_2D: MatrixReal4x4
+    ) -> List[Shapely]:
+        """Return a list of shapely geometries at the plane specified by normal and origin.
+
+        Parameters
+        ----------
+        normal : Coordinate
+            Vector defining the normal direction to the plane.
+        origin : Coordinate
+            Vector defining the plane origin.
+        to_2D : MatrixReal4x4
+            Transformation matrix to apply to resulting shapes.
+
+        Returns
+        -------
+        List[shapely.geometry.base.BaseGeometry]
+            List of 2D shapes that intersect plane.
+            For more details refer to
+            `Shapely's Documentaton <https://shapely.readthedocs.io/en/stable/project.html>`_.
+        """
+        return self.geometry.intersections_tilted_plane(
+            tuple(np.dot((normal[0], normal[1], normal[2], 0), self.transform)[:3]),
+            tuple(np.dot(self.inverse, (origin[0], origin[1], origin[2], 1))[:3]),
+            np.dot(to_2D, self.transform),
+        )
+
+    def inside(
+        self, x: np.ndarray[float], y: np.ndarray[float], z: np.ndarray[float]
+    ) -> np.ndarray[bool]:
+        """For input arrays ``x``, ``y``, ``z`` of arbitrary but identical shape, return an array
+        with the same shape which is ``True`` for every point in zip(x, y, z) that is inside the
+        volume of the :class:`Geometry`, and ``False`` otherwise.
+
+        Parameters
+        ----------
+        x : np.ndarray[float]
+            Array of point positions in x direction.
+        y : np.ndarray[float]
+            Array of point positions in y direction.
+        z : np.ndarray[float]
+            Array of point positions in z direction.
+
+        Returns
+        -------
+        np.ndarray[bool]
+            ``True`` for every point that is inside the geometry.
+        """
+        x = np.array(x)
+        y = np.array(y)
+        z = np.array(z)
+        xyz = np.dot(
+            self.inverse,
+            np.vstack((np.array(x).flat, np.array(y).flat, np.array(z).flat, np.ones(x.size))),
+        )
+        return self.geometry.inside(xyz[0], xyz[1], xyz[2]).reshape(x.shape)
+
+    def _volume(self, bounds: Bound) -> float:
+        """Returns object's volume within given bounds."""
+        (x0, y0, z0), (x1, y1, z1) = bounds
+        vertices = np.dot(
+            np.array(
+                [
+                    (x0, y0, z0, 1.0),
+                    (x0, y0, z1, 1.0),
+                    (x0, y1, z0, 1.0),
+                    (x0, y1, z1, 1.0),
+                    (x1, y0, z0, 1.0),
+                    (x1, y0, z1, 1.0),
+                    (x1, y1, z0, 1.0),
+                    (x1, y1, z1, 1.0),
+                ]
+            ),
+            self.inverse.T,
+        )
+        # NOTE (Lucas): Bounds are overestimated.
```

Comment:
in #1201 for example, we have some code checking if objects are 2D by comparing their bound mins and maxs, will overestimating the bounds be a problem for this?

Reply:
If the transformation doesn't involve the axis along which the geometry is 2D, it should not matter. For example: a 2D box with dimension 0 along z that is translated or scaled anywhere, or rotated around z will still have 0 bounds along z. If it is rotated around x or y, then that won't be true anymore.

Reply:
ok sounds good, thanks.